### PR TITLE
Creating Stages and Automation through DVC

### PR DESCRIPTION
**Pull request by daniel-chou-rainho**
To set up DVC and bind to my google drive cloud remote, do the following:

`pip install -r requirements.txt`
`dvc init`
`dvc remote add -d mygdrive gdrive://1XhzcyrUHVEerPQgE-_JjtkZvDfEihHtF`

Please then verify that you can fetch artifacts from DVC using `dvc pull`.

Then check if indeed all stages can be performed using `dvc repro`.
For now, I have not only included the `train.txt` file, but also a modified `train_sample.txt` file.
This file contains only 5% of the original training data, and is only meant to test if the automated pipeline works and gives sensible outputs. By default, the `dvc.yaml` file, which defines all stages, is now configured to use the sample training file, but this can easily be changed.

### CODE REVIEW

**Comment by erenaydoslu** [comment]
Could you share the file structure under /data? I'm getting an error when running `dvc repro` because it cannot find the paths

**Comment by daniel-chou-rainho** [comment]
image

`data` directory inside root folder.
and then 2 sub-directories, `raw` and `interim`.

**Comment by erenaydoslu** [comment]
image

I'm getting 0 trainable params, is this just a printing issue or is the model set incorrectly?

Other than that, everything looks good. Nicely done

**Comment by daniel-chou-rainho** [comment]
Hmm. Something is going wrong maybe, but am not sure.
I didn't fully run the last stage, because my laptop was just fanning like crazy, even just using the sample training dataset, which contains a lot less data. But at least it showed progress bars for stuff that was happening, so I assumed there were no errors in the code. Maybe I wrongly modified one of the files to output something that is not like before, not sure.
I would have to check the intermediary outputs as from running the original script and the stages here.

**Comment by erenaydoslu** [comment]
Something seems to be wrong in encoded_labels.txt as well. If I understand correctly, they should be 0 or 1 (legitimate or phishing). This is how it looks: 

image

Maybe encode_labels.py is encoding the urls, not the labels?

**Comment by daniel-chou-rainho** [comment]
Maybe I messed stuff up by trying to integrate batch processing.
Let me try revert the encode_labels part to the simple code from before and see if it works.

**Comment by daniel-chou-rainho** [comment]
Nevermind, that didn't help either.

**Comment by samuelribaric** [comment]
> image
> 
> I'm getting 0 trainable params, is this just a printing issue or is the model set incorrectly?
> 
> Other than that, everything looks good. Nicely done

Yeah I got the same thing when I tried, I think it happens because we do model.summary() before the model is trained (model.fit), looking into that now to confirm.

**Comment by daniel-chou-rainho** [comment]
I am restructuring things right now.
As it turns out some of the logic looks a bit too different from the original `main.py` file. I will try to re-create the stages to follow more closely the original `main.py` file.

**Comment by samuelribaric** [comment]
> I am restructuring things right now. As it turns out some of the logic looks a bit too different from the original `main.py` file. I will try to re-create the stages to follow more closely the original `main.py` file.

Yes, I had to change the files quite a bit to increase adaptability of the scripts, right now I have a finalized version that works for training as well as testing of the model. I also trained the model fully through Kaggle and added the whole model in /model/best_model.keras. If you want to process and encode a subset you can use the function `load_and_preprocess_data()` in `dataloader.py`.

**Comment by daniel-chou-rainho** [comment]
Sounds good. So everything works as expected now?

**Comment by daniel-chou-rainho** [comment]
One of the things I noticed was for example that the test and val datasets were not used.

**Comment by samuelribaric** [comment]
> One of the things I noticed was for example that the test and val datasets were not used.

Yes now the model performs as expected (slight better actually). 

image

Previously the val datasets was not used, if you take a look at `train.py` now you should see:
`    model.fit(x_train, y_train, batch_size=5000, epochs=30, validation_data=(x_val, y_val), callbacks=[checkpoint_callback])
`

**Comment by daniel-chou-rainho** [comment]
Will you commit your changes here?

**Comment by daniel-chou-rainho** [comment]
I made considerable changes so everything is more similar to the original `main.py` file.
As a consequence, the `encode_labels` stage now correctly outputs files with 0s and 1s.
However, just as before, my laptop cannot handle the final training stage, so if someone could check this, thanks.

**Comment by GiovanniLoureiro** [comment]
This seems to conflict with https://github.com/samuelribaric/REMLA-Project/pull/7, let's merge these two branches first before merging to main?

**Review by erenaydoslu — CHANGES_REQUESTED** [review]
Could you also add dvc to poetry requirements using `poetry add dvc`?

I run into issues running the code, maybe due to dependency issues. I did the following steps:
1. `poetry add dvc`
2. `poetry add dvc_gdrive`
3. `poetry install`
4. `poetry shell`
5. `dvc repro`

Check readme.md for poetry installation instructions.

The error I receive when running the code:
image

**Review by erenaydoslu — APPROVED** [review]
